### PR TITLE
feat: player external play

### PR DESCRIPTION
**Pull request by tymmesyde**
## Related PRs and Issues

https://github.com/Stremio/stremio-core/pull/421
https://github.com/Stremio/stremio-core/pull/427
~https://github.com/Stremio/stremio-core-web/pull/40~ closed in favour of https://github.com/Stremio/stremio-core-web/pull/41

## Feature
Add options to open streams in external players

image

### CODE REVIEW

**Review by elpiel — COMMENTED** [review]
~Code looks good to me, however, testing under **Firefox** gives me a few problems:~

~https://user-images.githubusercontent.com/8925621/228165968-f0ec296b-ba93-4035-a138-3a70db8c7c2e.mp4~


1. I couldn't find a single stream that is supported :sweat_smile: 
~2. Clicking on the 3 dots menu, I get multiple errors in the console and I get redirected to:
[link](https://localhost:8080/#/player/eAEBSAm39nsiaW5mb0hhc2giOiI1MmE4YzU0OTcwY2UwNjY0Nzg2YmU5YzZmZWY0YjE5YTZiMmI1NzgzIiwiZmlsZUlkeCI6NywiYW5ub3VuY2UiOlsidHJhY2tlcjpodHRwOi8vYnQyLnQtcnUub3JnL2Fubj9tYWduZXQiLCJkaHQ6NTJhOGM1NDk3MGNlMDY2NDc4NmJlOWM2ZmVmNGIxOWE2YjJiNTc4MyIsInRyYWNrZXI6dWRwOi8vdHJhY2tlci5vcGVudHJhY2tyLm9yZzoxMzM3L2Fubm91bmNlIiwidHJhY2tlcjp1ZHA6Ly9vcGVudHJhY2tlci5pMnAucm9ja3M6Njk2OS9hbm5vdW5jZSIsInRyYWNrZXI6aHR0cHM6Ly9vcGVudHJhY2tlci5pMnAucm9ja3M6NDQzL2Fubm91bmNlIiwidHJhY2tlcjp1ZHA6Ly90cmFja2VyLm9wZW5iaXR0b3JyZW50LmNvbTo2OTY5L2Fubm91bmNlIiwidHJhY2tlcjpodHRwOi8vdHJhY2tlci5vcGVuYml0dG9ycmVudC5jb206ODAvYW5ub3VuY2UiLCJ0cmFja2VyOnVkcDovLzkucmFyYmcuY29tOjI4MTAvYW5ub3VuY2UiLCJ0cmFja2VyOnVkcDovL29wZW4uZGVtb25paS5jb206MTMzNy9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vZXhvZHVzLmRlc3luYy5jb206Njk2OS9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vb3Blbi5zdGVhbHRoLnNpOjgwL2Fubm91bmNlIiwidHJhY2tlcjp1ZHA6Ly90cmFja2VyLnRvcnJlbnQuZXUub3JnOjQ1MS9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vaXB2NC50cmFja2VyLmhhcnJ5Lmx1OjgwL2Fubm91bmNlIiwidHJhY2tlcjpodHRwczovL3RyYWNrZXIyLmN0aXguY246NDQzL2Fubm91bmNlIiwidHJhY2tlcjpodHRwczovL3RyYWNrZXIxLjUyMC5qcDo0NDMvYW5ub3VuY2UiLCJ0cmFja2VyOmh0dHBzOi8vdHJhY2tlci50YW1lcnN1bmlvbi5vcmc6NDQzL2Fubm91bmNlIiwidHJhY2tlcjpodHRwOi8vdnBzMDIubmV0Lm9yZWwucnU6ODAvYW5ub3VuY2UiLCJ0cmFja2VyOnVkcDovL3RyYWNrZXIubW9la2luZy5tZTo2OTY5L2Fubm91bmNlIiwidHJhY2tlcjp1ZHA6Ly90cmFja2VyMS5idC5tb2Fjay5jby5rcjo4MC9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vdHJhY2tlci11ZHAuZ2JpdHQuaW5mbzo4MC9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vZXhwbG9kaWUub3JnOjY5NjkvYW5ub3VuY2UiLCJ0cmFja2VyOmh0dHBzOi8vdHJhY2tlci5nYml0dC5pbmZvOjQ0My9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vNDcuaXAtNTEtNjgtMTk5LmV1OjY5NjkvYW5ub3VuY2UiLCJ0cmFja2VyOnVkcDovLzkucmFyYmcubWU6Mjk0MCIsInRyYWNrZXI6dWRwOi8vOS5yYXJiZy50bzoyODIwIiwidHJhY2tlcjp1ZHA6Ly9vcGVudG9yLm9yZzoyNzEwL2Fubm91bmNlIiwidHJhY2tlcjp1ZHA6Ly9yZXRyYWNrZXIubGFudGEtbmV0LnJ1OjI3MTAvYW5ub3VuY2UiLCJ0cmFja2VyOnVkcDovL3RyYWNrZXIuY3liZXJpYS5pczo2OTY5L2Fubm91bmNlIiwidHJhY2tlcjp1ZHA6Ly90cmFja2VyLmRsZXIub3JnOjY5NjkvYW5ub3VuY2UiLCJ0cmFja2VyOnVkcDovL3RyYWNrZXIuZHMuaXM6Njk2OS9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vdHJhY2tlci5pbnRlcm5ldHdhcnJpb3JzLm5ldDoxMzM3IiwidHJhY2tlcjp1ZHA6Ly90cmFja2VyLnRpbnktdnBzLmNvbTo2OTY5L2Fubm91bmNlIiwidHJhY2tlcjp1ZHA6Ly92YWxha2FzLnJvbGxvLmRuc2Fici5jb206MjcxMC9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vd3d3LnRvcnJlbnQuZXUub3JnOjQ1MS9hbm5vdW5jZSIsInRyYWNrZXI6dWRwOi8vb3BlbnRvci5vcmc6MjcxMCIsInRyYWNrZXI6aHR0cDovL3JldHJhY2tlci5sb2NhbC9hbm5vdW5jZSIsInRyYWNrZXI6aHR0cDovL2J0LnQtcnUub3JnL2Fubj9tYWduZXQiLCJ0cmFja2VyOmh0dHA6Ly9idDMudC1ydS5vcmcvYW5uP21hZ25ldCIsInRyYWNrZXI6aHR0cDovL2J0NC50LXJ1Lm9yZy9hbm4%2FbWFnbmV0Il0sIm5hbWUiOiJUb3JyZW50aW9cbjEwODBwIiwiZGVzY3JpcHRpb24iOiLQlNC20LXQuiDQoNCw0LnQsNC9IC8gSmFjayBSeWFuIC8g0KHQtdC30L7QvTogMyAvINCh0LXRgNC40Lg6IDEtOCDQuNC3IDggKNCv0L3QvSDQotC10YDQvdC10YApIFsyMDIyINCh0KjQkCDQsdC%2B0LXQstC40Log0YLRgNC40LvQu9C10YAg0LTRgNCw0LzQsCBXRUItREwgMTA4MHBdIE1WTyAoTG9zdEZpbG0gSERyZXprYSBTdHVkaW8pICsgT3JpZ2luYWwgKyBTdWIgKFJ1cyBFbmcpXG5Ub20uQ2xhbmN5cy5KYWNrLlJ5YW4uUzAzRTA4LldFQkRMLjEwODBwLlJHenNSdXRyYWNrZXIubWt2XG7wn5GkIDMg8J%2BSviAyLjUzIEdCIOKame%2B4jyBSdXRyYWNrZXJcbvCfh6zwn4enIC8g8J%2BHt%2FCfh7oiLCJiZWhhdmlvckhpbnRzIjp7ImJpbmdlR3JvdXAiOiJ0b3JyZW50aW98NTJhOGM1NDk3MGNlMDY2NDc4NmJlOWM2ZmVmNGIxOWE2YjJiNTc4MyJ9fUxyVoE%3D/https%3A%2F%2Ftorrentio.strem.fun%2Fmanifest.json/https%3A%2F%2Fv3-cinemeta.strem.io%2Fmanifest.json/series/tt5057054/tt5057054%3A3%3A8)~

~Errors:~

```
Uncaught TypeError: streamingServer.playbackDevices is undefined
The above error occurred in the <Player> component
Uncaught TypeError: core.transport is null
The above error occurred in the <ServicesToaster> component
```

The issues above were caused by not running `npm install` before testing the PR.
Sorry about that!

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
         "@babel/runtime": "7.16.0",
         "@sentry/browser": "6.13.3",
         "@stremio/stremio-colors": "5.0.1",
-        "@stremio/stremio-core-web": "0.44.11",
+        "@stremio/stremio-core-web": "0.44.12",
```
Comment (elpiel) [inline package.json]: - [ ] Please update to `0.44.13`

Comment (elpiel) [inline package.json]:
```suggestion
        "@stremio/stremio-core-web": "0.44.13",
```

File: src/routes/Player/OptionsMenu/OptionsMenu.js [hunk 2]
```diff
@@ -38,55 +38,78 @@ const OptionsMenu = ({ className, stream }) => {
...
+                    />
+                    :
+                    null
+            }
+            {
+                streamingUrl && externalDevices.map(({ id, name }) => (
+                    <Option
+                        key={id}
+                        icon={'ic_vlc'}
+                        label={`Play in ${name}`}
```
Comment (elpiel) [inline src/routes/Player/OptionsMenu/OptionsMenu.js:98]: - [ ] Should be translated in future PR

File: src/routes/Player/OptionsMenu/OptionsMenu.js [hunk 3]
```diff
@@ -38,55 +38,78 @@ const OptionsMenu = ({ className, stream }) => {
...
-            </Button>
-            {/* <Button className={classnames(styles['option-container'], { 'disabled': stream === null })} disabled={stream === null} onClick={onExternalPlayerButtonClick}>
-                <Icon className={styles['icon']} icon={'ic_vlc'} />
-                <div className={styles['label']}>Play in External Player</div>
-            </Button> */}
+            {
+                streamingUrl || downloadUrl ?
+                    <Option
+                        icon={'ic_link'}
+                        label={'Copy Stream Link'}
```
Comment (elpiel) [inline src/routes/Player/OptionsMenu/OptionsMenu.js:75]: - [ ] Should be translated in future PR

File: src/routes/Player/OptionsMenu/OptionsMenu.js [hunk 4]
```diff
@@ -38,55 +38,78 @@ const OptionsMenu = ({ className, stream }) => {
                     toast.show({
                         type: 'error',
                         title: 'Error',
-                        message: `Failed to copy stream link: ${streamUrl}`,
+                        message: `Failed to copy stream link: ${streamingUrl || downloadUrl}`,
```
Comment (elpiel) [inline src/routes/Player/OptionsMenu/OptionsMenu.js:41]: - [ ] Should be translated in future PR?!

File: src/App/ServicesToaster.js [hunk 5]
```diff
@@ -50,6 +50,14 @@ const ServicesToaster = () => {
                     });
                     break;
                 }
+                case 'PlayingOnDevice': {
+                    toast.show({
+                        type: 'success',
+                        title: `Stream opened in ${args.device}`,
```
Comment (elpiel) [inline src/App/ServicesToaster.js:56]: - [ ] Should be translated in future PR

**Review by elpiel — APPROVED** [review]
LGTM and tested with VLC.

File: src/routes/Player/OptionsMenu/Option/Option.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
+// Copyright (C) 2017-2022 Smart code 203358507
+
+const React = require('react');
+const PropTypes = require('prop-types');
+const classnames = require('classnames');
+const Icon = require('@stremio/stremio-icons/dom');
+const { Button } = require('stremio/common');
+const styles = require('./styles');
+
+const Option = ({ icon, label, deviceId, disabled, onClick }) => {
```
Comment (elpiel) [inline src/routes/Player/OptionsMenu/Option/Option.js:10]: - [x] ~We should add some docs for parameters, e.g.:~ As discussed this is already very clear and will be much clearer if we switch to Typescript
  - `onClick` should take `deviceId` parameter
  - `label` should be a translated string as the component doesn't translates it
  - etc.

**Review by nklhtv — APPROVED** [review]

**Pull request merged**